### PR TITLE
feat: update l1fee calculation

### PR DESCRIPTION
**Pull request by 0xmountaintop**
## 1. Purpose or design rationale of this PR

update L1fee calculation. now should include signature

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [x] Yes

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204700917052368

### CODE REVIEW

File: core/state_transition.go [hunk 1]
```diff
@@ -196,19 +192,19 @@ func NewStateTransition(evm *vm.EVM, msg Message, gp *GasPool) *StateTransition
...
 // against the old state within the environment.
 //
-// ApplyMessage returns the bytes returned by any EVM execution (if it took place),
+// ApplyMessageAndL1DataFee returns the bytes returned by any EVM execution (if it took place),
 // the gas used (which includes gas refunds) and an error if it failed. An error always
 // indicates a core error meaning that the message would always fail for that particular
 // state and would never be accepted within a block.
-func ApplyMessage(evm *vm.EVM, msg Message, gp *GasPool) (*ExecutionResult, error) {
-	return NewStateTransition(evm, msg, gp).TransitionDb()
+func ApplyMessageAndL1DataFee(evm *vm.EVM, msg Message, gp *GasPool, l1DataFee *big.Int) (*ExecutionResult, error) {
```
Comment (icemelon) [inline core/state_transition.go]: We can still keep the same function name `ApplyMessage`.

File: core/state_transition.go [hunk 2]
```diff
@@ -44,8 +43,10 @@ The state transitioning model does all the necessary work to work out a valid ne
 3) Create a new state object if the recipient is \0*32
 4) Value transfer
 == If contract creation ==
-  4a) Attempt to run transaction data
-  4b) If valid, use result as code for the new state object
+
+	4a) Attempt to run transaction data
+	4b) If valid, use result as code for the new state object
+
```
Comment (Thegaram) [inline core/state_transition.go]: Should we undo these format changes? (Also a couple of other places below.)

File: core/state_transition.go [hunk 3]
```diff
@@ -64,7 +65,7 @@ type StateTransition struct {
 	evm        *vm.EVM
 
 	// l1 rollup fee
-	l1Fee *big.Int
+	l1Fee *big.Int // TODO: rename to l1DataFee
```
Comment (Thegaram) [inline core/state_transition.go]: Do you plan to rename it in another (refactor) PR?

File: rollup/fees/rollup_fee.go [hunk 4]
```diff
@@ -48,59 +44,82 @@ type StateDB interface {
...
-		return types.NewContractCreation(
-			msg.Nonce(),
-			msg.Value(),
-			msg.Gas(),
-			msg.GasPrice(),
-			msg.Data(),
-		)
+// asUnsignedTx turns a Message into a types.Transaction
+func asUnsignedTx(msg Message, baseFee, chainID *big.Int) *types.Transaction {
+	if msg.AccessList() == nil {
```
Comment (Thegaram) [inline rollup/fees/rollup_fee.go]: Is it possible that `msg.AccessList() == nil` but `baseFee != nil`?

File: rollup/fees/rollup_fee.go [hunk 5]
```diff
@@ -150,30 +169,8 @@ func mulAndScale(x *big.Int, y *big.Int, precision *big.Int) *big.Int {
...
-		tx.Nonce(),
-		*tx.To(),
-		tx.Value(),
-		tx.Gas(),
-		tx.GasPrice(),
-		tx.Data(),
-	)
-}
-
 func CalculateFees(tx *types.Transaction, state StateDB) (*big.Int, *big.Int, *big.Int, error) {
```
Comment (Thegaram) [inline rollup/fees/rollup_fee.go]: ~Add~ Are `l2Fee` and `fee` used anywhere?

File: eth/tracers/api.go [hunk 6]
```diff
@@ -283,7 +283,14 @@ func (api *API) traceChain(ctx context.Context, start, end *types.Block, config
 						TxIndex:   i,
 						TxHash:    tx.Hash(),
 					}
-					res, err := api.traceTx(localctx, msg, txctx, blockCtx, task.statedb, config)
+
+					l1DataFee, _, _, err := fees.CalculateFees(tx, task.statedb)
+					if err != nil {
+						log.Warn("CalculateFees failed", "hash", tx.Hash(), "block", task.block.NumberU64(), "err", err)
+						break
```
Comment (Thegaram) [inline eth/tracers/api.go:292]: Why not return error here?

File: eth/tracers/api.go [hunk 7]
```diff
@@ -534,7 +541,14 @@ func (api *API) IntermediateRoots(ctx context.Context, hash common.Hash, config
 			vmenv     = vm.NewEVM(vmctx, txContext, statedb, chainConfig, vm.Config{})
 		)
 		statedb.Prepare(tx.Hash(), i)
-		if _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas())); err != nil {
+
+		l1DataFee, _, _, err := fees.CalculateFees(tx, statedb)
+		if err != nil {
+			log.Warn("Tracing intermediate roots did not complete due to fees.CalculateFees", "txindex", i, "txhash", tx.Hash(), "err", err)
+			return roots, nil
```
Comment (Thegaram) [inline eth/tracers/api.go]: ditto

File: core/state_transition.go (hunk 3 above)
Comment (icemelon) [reply core/state_transition.go]: better to change it in the this PR?

File: core/state_transition.go [hunk 8]
```diff
@@ -307,13 +303,13 @@ func (st *StateTransition) preCheck() error {
...
 // returning the evm execution result with following fields.
 //
-// - used gas:
-//      total gas used (including gas being refunded)
-// - returndata:
-//      the returned data from evm
-// - concrete execution error:
-//      various **EVM** error which aborts the execution,
-//      e.g. ErrOutOfGas, ErrExecutionReverted
+//   - used gas:
```
Comment (icemelon) [inline core/state_transition.go]: revert the format change?

File: rollup/fees/rollup_fee.go [hunk 9]
```diff
@@ -48,59 +44,82 @@ type StateDB interface {
...
 }
 
-// CalculateL1MsgFee computes the L1 portion of the fee given
-// a Message and a StateDB
-// Reference: https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/fees/rollup_fee.go
-func CalculateL1MsgFee(msg Message, state StateDB) (*big.Int, error) {
-	tx := asTransaction(msg)
+func EstimateL1DataFeeForMessage(msg Message, baseFee, chainID *big.Int, signer types.Signer, state StateDB) (*big.Int, error) {
+	unsigned := asUnsignedTx(msg, baseFee, chainID)
+	tx, err := unsigned.WithSignature(signer, bytes.Repeat([]byte{0xff}, crypto.SignatureLength))
```
Comment (icemelon) [inline rollup/fees/rollup_fee.go]: Should we set `v` to 1 since the value could have impact on the `V` value in the EIP155 tx?

File: rollup/fees/rollup_fee.go (hunk 4 above)
Comment (icemelon) [reply rollup/fees/rollup_fee.go]: Currently not since EIP-1559 tx also has an access list.

File: rollup/fees/rollup_fee.go (hunk 5 above)
Comment (icemelon) [reply rollup/fees/rollup_fee.go]: you mean rename the func to `CalculateL1DataFees`?

File: eth/state_accessor.go [hunk 10]
```diff
@@ -36,15 +37,15 @@ import (
...
-//        it would be preferrable to start from a fresh state, if we have it on disk.
+//   - block: The block for which we want the state (== state at the stateRoot of the parent)
+//   - reexec: The maximum number of blocks to reprocess trying to obtain the desired state
+//   - base: If the caller is tracing multiple blocks, the caller can provide the parent state
+//     continuously from the callsite.
+//   - checklive: if true, then the live 'blockchain' state database is used. If the caller want to
+//     perform Commit or other 'save-to-disk' changes, this should be set to false to avoid
+//     storing trash persistently
+//   - preferDisk: this arg can be used by the caller to signal that even though the 'base' is provided,
+//     it would be preferrable to start from a fresh state, if we have it on disk.
```
Comment (icemelon) [inline eth/state_accessor.go]: revert the formating change

File: rollup/fees/rollup_fee.go (hunk 5 above)
Comment (icemelon) [reply rollup/fees/rollup_fee.go]: For `L1MessageTx` tx, we should directly return 0 in the case of tracing such tx.

File: internal/ethapi/api.go [hunk 11]
```diff
@@ -741,10 +741,10 @@ func (s *PublicBlockChainAPI) GetHeaderByHash(ctx context.Context, hash common.H
 }
 
 // GetBlockByNumber returns the requested canonical block.
-// * When blockNr is -1 the chain head is returned.
-// * When blockNr is -2 the pending chain head is returned.
-// * When fullTx is true all transactions in the block are returned, otherwise
-//   only the transaction hash is returned.
+//   - When blockNr is -1 the chain head is returned.
```
Comment (icemelon) [inline internal/ethapi/api.go]: revert formating

File: core/state_transition.go (hunk 2 above)
Comment (0xmountaintop) [reply core/state_transition.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/c7a1066892649ebecfd3a47082b0e21fc1aa85bb

File: core/state_transition.go (hunk 8 above)
Comment (0xmountaintop) [reply core/state_transition.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/c7a1066892649ebecfd3a47082b0e21fc1aa85bb

File: eth/state_accessor.go (hunk 10 above)
Comment (0xmountaintop) [reply eth/state_accessor.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/c7a1066892649ebecfd3a47082b0e21fc1aa85bb

File: internal/ethapi/api.go (hunk 11 above)
Comment (0xmountaintop) [reply internal/ethapi/api.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/c7a1066892649ebecfd3a47082b0e21fc1aa85bb

File: core/state_transition.go (hunk 3 above)
Comment (0xmountaintop) [reply core/state_transition.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/6e32db016cb44ce8435a911553b6e032921e80d0

File: rollup/fees/rollup_fee.go (hunk 5 above)
Comment (0xmountaintop) [reply rollup/fees/rollup_fee.go]: > you mean rename the func to `CalculateL1DataFees`?

We already have a `CalculateL1DataFee`. I think peter means "**Are** `l2Fee` and `fee` used anywhere?"

`l2Fee` is used in `fees.VerifyFee()`

---

> For L1MessageTx tx, we should directly return 0 in the case of tracing such tx.

~~So, should I base on `relay-l1-messages` branch? Currently `L1MessageTx` is not there on `develop` yet.~~

~~Or I can add this logic shortly, after https://github.com/scroll-tech/go-ethereum/pull/350 is merged (which seems will happen soon).~~

Comment (icemelon) [reply rollup/fees/rollup_fee.go]: Maybe rename this function to `CalculateTxL1DataFee` to be more precise? alternatively, we could make the current `CalculateL1DataFee` to an internal function eg. `calculateFee`, and make this function `CalculateL1DataFee`

File: core/state_transition.go [hunk 12]
```diff
@@ -225,9 +218,9 @@ func (st *StateTransition) buyGas() error {
 	mgval = mgval.Mul(mgval, st.gasPrice)
 
 	if st.evm.ChainConfig().Scroll.FeeVaultEnabled() {
-		// always add l1fee, because all tx are L2-to-L1 ATM
-		log.Debug("Adding L1 fee", "l1_fee", st.l1Fee)
-		mgval = mgval.Add(mgval, st.l1Fee)
+		// always add l1DataFee, because all tx are L2-to-L1 ATM
```
Comment (icemelon) [inline core/state_transition.go]: We should remove this comment or revise it because the `l1DataFee` for L1MessageTx should be 0

File: core/state_transition.go [hunk 13]
```diff
@@ -236,8 +229,8 @@ func (st *StateTransition) buyGas() error {
 		balanceCheck = balanceCheck.Mul(balanceCheck, st.gasFeeCap)
 		balanceCheck.Add(balanceCheck, st.value)
 		if st.evm.ChainConfig().Scroll.FeeVaultEnabled() {
-			// always add l1fee, because all tx are L2-to-L1 ATM
-			balanceCheck.Add(balanceCheck, st.l1Fee)
+			// always add l1DataFee, because all tx are L2-to-L1 ATM
```
Comment (icemelon) [inline core/state_transition.go]: ditto

File: rollup/fees/rollup_fee.go (hunk 5 above)
Comment (0xmountaintop) [reply rollup/fees/rollup_fee.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/aac08fbfd05fa8a94b69a90fa8b9d83d165d1b3d

File: core/state_transition.go (hunk 12 above)
Comment (0xmountaintop) [reply core/state_transition.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/9e223e9db85603a82aac0a859c28ffa4045cbfeb

File: core/state_transition.go (hunk 13 above)
Comment (0xmountaintop) [reply core/state_transition.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/9e223e9db85603a82aac0a859c28ffa4045cbfeb

File: eth/tracers/api.go (hunk 7 above)
Comment (0xmountaintop) [reply eth/tracers/api.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/71b775d03efd523f32b9d543be379f4b66b1cbe2

File: eth/tracers/api.go (hunk 6 above)
Comment (0xmountaintop) [reply eth/tracers/api.go:292]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/71b775d03efd523f32b9d543be379f4b66b1cbe2

File: eth/tracers/api.go (hunk 7 above)
Comment (0xmountaintop) [reply eth/tracers/api.go]: i was thinking about this, and thought I'd better "intentionally don't return the error" as what `ApplyMessage` does below.

but after rethinking about it i agree we should return error

File: core/state_transition.go [hunk 14]
```diff
@@ -225,9 +218,12 @@ func (st *StateTransition) buyGas() error {
 	mgval = mgval.Mul(mgval, st.gasPrice)
 
 	if st.evm.ChainConfig().Scroll.FeeVaultEnabled() {
```
Comment (icemelon) [inline core/state_transition.go:220]: One question about this condition check. If the fee vault is disabled in the future, we still need to charge the L1 data fee, right? I wonder if we should use a different flag, e.g., `Scroll.Enabled()`, to represent if the Scroll protocol is enabled?

File: eth/tracers/api.go [hunk 15]
```diff
@@ -740,15 +768,16 @@ func (api *API) standardTraceBlockToFile(ctx context.Context, block *types.Block
 		// Execute the transaction and flush any traces to disk
 		vmenv := vm.NewEVM(vmctx, txContext, statedb, chainConfig, vmConf)
 		statedb.Prepare(tx.Hash(), i)
-		_, err = core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+		l1DataFee, err1 := fees.CalculateL1DataFee(tx, statedb)
```
Comment (icemelon) [inline eth/tracers/api.go]: we should return early here if `CalculateL1DataFee` runs into an error

File: internal/ethapi/api.go [hunk 16]
```diff
@@ -907,6 +907,7 @@ func newRPCBalance(balance *big.Int) **hexutil.Big {
 	return &rpcBalance
 }
 
+// TODO: rename this API to EstimateL1MsgFee?
```
Comment (icemelon) [inline internal/ethapi/api.go]: yes, agree that changing to `EstimateL1MsgFee` is better.

File: core/state_transition.go (hunk 14 above)
Comment (0xmountaintop) [reply core/state_transition.go:220]: Peter and I think Scroll protocol will always be enabled, for example we always use zktrie.
we just need to clean up the codes (remove those `Enabled()` check)

File: internal/ethapi/api.go (hunk 16 above)
Comment (0xmountaintop) [reply internal/ethapi/api.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/b34ec91bc5f4806c3c774f766f20905081d8f036

File: eth/tracers/api.go (hunk 15 above)
Comment (0xmountaintop) [reply eth/tracers/api.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/f7d3b9d02fcf56981aaa56a60389fe48120b4f07

File: core/state_transition.go (hunk 14 above)
Comment (icemelon) [reply core/state_transition.go:220]: In that case, we should always charge the L1 data fee, right?

File: eth/tracers/api.go [hunk 17]
```diff
@@ -740,7 +768,18 @@ func (api *API) standardTraceBlockToFile(ctx context.Context, block *types.Block
...
 		vmenv := vm.NewEVM(vmctx, txContext, statedb, chainConfig, vmConf)
 		statedb.Prepare(tx.Hash(), i)
-		_, err = core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+		l1DataFee, err := fees.CalculateL1DataFee(tx, statedb)
+		if err != nil {
+			if writer != nil {
+				writer.Flush()
+			}
+			if dump != nil {
+				dump.Close()
```
Comment (icemelon) [inline eth/tracers/api.go]: print a log info similar to line 788

**Review by icemelon — DISMISSED** [review]
a minor comment. otherwise lgtm

File: rollup/fees/rollup_fee.go (hunk 4 above)
Comment (Thegaram) [reply rollup/fees/rollup_fee.go]: Yeah, but if the access list is empty, it might be `nil`.

File: core/state_transition.go (hunk 14 above)
Comment (Thegaram) [reply core/state_transition.go:220]: Yes, we'll always charge the L1 data fee, except for L1 messages (but we don't call `buyGas()` for those). I agree with Haoyu that we can clean up these flags in a separate PR.

File: core/state_transition.go [hunk 18]
```diff
@@ -236,8 +232,11 @@ func (st *StateTransition) buyGas() error {
 		balanceCheck = balanceCheck.Mul(balanceCheck, st.gasFeeCap)
 		balanceCheck.Add(balanceCheck, st.value)
 		if st.evm.ChainConfig().Scroll.FeeVaultEnabled() {
-			// always add l1fee, because all tx are L2-to-L1 ATM
-			balanceCheck.Add(balanceCheck, st.l1Fee)
+			// should be fine to add st.l1DataFee even without `L1MessageTx` check, since L1MessageTx will come with 0 l1DataFee,
+			// but double check to make sure
+			if !st.msg.IsL1MessageTx() {
```
Comment (Thegaram) [inline core/state_transition.go:237]: We short-circuit in `preCheck()` for `L1MessageTx` so we won't call `buyGas()` for it. But I don't mind leaving some extra checks here.

File: core/types/receipt.go [hunk 19]
```diff
@@ -75,7 +75,7 @@ type Receipt struct {
 	ReturnValue []byte `json:"returnValue,omitempty"`
 
 	// Scroll rollup
-	L1Fee *big.Int `json:"l1Fee,omitempty"`
+	L1DataFee *big.Int `json:"l1DataFee,omitempty"`
```
Comment (Thegaram) [inline core/types/receipt.go]: This is a breaking change. And it would require changes in `web3go` and `rpc-gateway`, too. I lean towards leaving it as `"l1Fee"` but let's discuss.

File: eth/tracers/api.go [hunk 20]
```diff
@@ -740,7 +768,19 @@ func (api *API) standardTraceBlockToFile(ctx context.Context, block *types.Block
...
+				log.Info("Wrote standard trace due to CalculateL1DataFee error", "file", dump.Name(), "err", err)
+			}
+			return dumps, err
+		}
+
+		_, err = core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()), l1DataFee)
```
Comment (Thegaram) [inline eth/tracers/api.go]: This could be cleaner. The `writer`, `dump`, and `err` check do not need to be duplicated.

```go
l1DataFee, err := fees.CalculateL1DataFee(tx, statedb)
if err == nil {
    _, err = core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()), l1DataFee)
}
if writer != nil {
// ...
```

File: core/types/l2trace.go [hunk 21]
```diff
@@ -44,7 +44,7 @@ type StorageTrace struct {
 // while replaying a transaction in debug mode as well as transaction
 // execution status, the amount of gas used and the return value
 type ExecutionResult struct {
-	L1Fee       uint64 `json:"l1Fee,omitempty"`
+	L1DataFee   uint64 `json:"l1DataFee,omitempty"`
```
Comment (Thegaram) [inline core/types/l2trace.go]: Not related to this PR but I just noticed that this should be big int.

File: internal/ethapi/api.go [hunk 22]
```diff
@@ -1052,13 +1060,13 @@ func (s *PublicBlockChainAPI) Call(ctx context.Context, args TransactionArgs, bl
 	_, isOverrideSet := (*overrides)[args.from()]
 
 	if isGasPriceZero && !isOverrideSet {
-		l1Fee, err := CalculateL1MsgFee(ctx, s.b, args, blockNrOrHash, overrides, s.b.RPCEVMTimeout(), s.b.RPCGasCap(), s.b.ChainConfig())
+		l1DataFee, err := EstimateL1MsgFee(ctx, s.b, args, blockNrOrHash, overrides, s.b.RPCEVMTimeout(), s.b.RPCGasCap(), s.b.ChainConfig())
```
Comment (Thegaram) [inline internal/ethapi/api.go:1063]: We call `fees.EstimateL1DataFeeForMessage` twice per call, here and in `DoCall`. But I think it's acceptable.

File: internal/ethapi/api.go [hunk 23]
```diff
@@ -1669,7 +1682,7 @@ func (s *PublicTransactionPoolAPI) GetTransactionReceipt(ctx context.Context, ha
 		"logs":              receipt.Logs,
 		"logsBloom":         receipt.Bloom,
 		"type":              hexutil.Uint(tx.Type()),
-		"l1Fee":             (*hexutil.Big)(receipt.L1Fee),
+		"l1DataFee":         (*hexutil.Big)(receipt.L1DataFee),
```
Comment (Thegaram) [inline internal/ethapi/api.go]: Note: If we change it in `Receipt`, we need to change it here, too.

File: rollup/fees/rollup_fee.go [hunk 24]
```diff
@@ -48,59 +45,87 @@ type StateDB interface {
...
-// Reference: https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/fees/rollup_fee.go
-func CalculateL1MsgFee(msg Message, state StateDB) (*big.Int, error) {
-	tx := asTransaction(msg)
+func EstimateL1DataFeeForMessage(msg Message, baseFee, chainID *big.Int, signer types.Signer, state StateDB) (*big.Int, error) {
+	if msg.IsL1MessageTx() {
+		return big.NewInt(0), nil
+	}
+
+	unsigned := asUnsignedTx(msg, baseFee, chainID)
+	// with v=1
```
Comment (Thegaram) [inline rollup/fees/rollup_fee.go:54]: Why `v = 0x01` and not `v = 0xff` or `v = 0xffffff`?

File: rollup/fees/rollup_fee.go [hunk 25]
```diff
@@ -48,59 +45,87 @@ type StateDB interface {
...
-			msg.GasPrice(),
-			msg.Data(),
-		)
+	l1DataFee := calculateEncodedL1DataFee(raw, overhead, l1BaseFee, scalar)
+	return l1DataFee, nil
+}
+
+// asUnsignedTx turns a Message into a types.Transaction
+func asUnsignedTx(msg Message, baseFee, chainID *big.Int) *types.Transaction {
+	if msg.AccessList() == nil {
```
Comment (Thegaram) [inline rollup/fees/rollup_fee.go]:
```suggestion
	if msg.AccessList() == nil && baseFee == nil {
```

File: rollup/fees/rollup_fee.go [hunk 26]
```diff
@@ -206,9 +214,9 @@ func VerifyFee(signer types.Signer, tx *types.Transaction, state StateDB) error
 		return errors.New("invalid transaction: insufficient funds for gas * price + value")
 	}
 
-	cost = cost.Add(cost, l1Fee)
+	cost = cost.Add(cost, l1DataFee)
 	if balance.Cmp(cost) < 0 {
-		return errors.New("invalid transaction: insufficient funds for l1fee + gas * price + value")
+		return errors.New("invalid transaction: insufficient funds for l1DataFee + gas * price + value")
```
Comment (Thegaram) [inline rollup/fees/rollup_fee.go]: This is also a breaking change and some projects like ChainLink might already rely on this error message. We could change it any notify them but I think it's easier to keep it as `l1fee`.

Comment (0xmountaintop) [reply rollup/fees/rollup_fee.go]: I will revert this one

File: core/types/l2trace.go (hunk 21 above)
Comment (Thegaram) [reply core/types/l2trace.go]: Fixed in https://github.com/scroll-tech/go-ethereum/pull/360.

File: core/types/receipt.go (hunk 19 above)
Comment (0xmountaintop) [reply core/types/receipt.go]: cc @icemelon

File: rollup/fees/rollup_fee.go (hunk 24 above)
Comment (0xmountaintop) [reply rollup/fees/rollup_fee.go:54]: v will only be 0 or 1, see core/types/transaction_signing.go

Using 1 to estimate seems more accurate

Comment (Thegaram) [reply rollup/fees/rollup_fee.go:54]: I see. The chainId is added to `v` (for legacy tx) in the `WithSignature` call.

File: rollup/fees/rollup_fee.go (hunk 26 above)
Comment (0xmountaintop) [reply rollup/fees/rollup_fee.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/2753504c3518df5f66ecad89caf7e967d02dbb4c

File: eth/tracers/api.go (hunk 20 above)
Comment (0xmountaintop) [reply eth/tracers/api.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/2753504c3518df5f66ecad89caf7e967d02dbb4c

File: rollup/fees/rollup_fee.go (hunk 25 above)
Comment (0xmountaintop) [reply rollup/fees/rollup_fee.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/ab6c5c4d2c38375ea31cccba0e065df4c727de18

File: core/types/receipt.go (hunk 19 above)
Comment (Thegaram) [reply core/types/receipt.go]: Advantage of keeping `"l1Fee"`: compatibility with Optimism, maybe getting one step closer to an industry-wide standard.

Comment (icemelon) [reply core/types/receipt.go]: Agree. Let's keep it `l1Fee` in the receipt.

Comment (0xmountaintop) [reply core/types/receipt.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/cad96570acda00d701d09f9c5d4f04805bba60bb

File: core/rawdb/accessors_chain.go [hunk 27]
```diff
@@ -623,7 +623,7 @@ type storedReceiptRLP struct {
 	PostStateOrStatus []byte
 	CumulativeGasUsed uint64
 	Logs              []*types.LogForStorage
-	L1Fee             *big.Int
+	L1DataFee         *big.Int
```
Comment (Thegaram) [inline core/rawdb/accessors_chain.go]:
```suggestion
	L1Fee         *big.Int
```

This is not used here, but still better to keep it consistent with `core/types/receipt.go`.

File: core/state_prefetcher.go [hunk 28]
```diff
@@ -68,7 +70,13 @@ func (p *statePrefetcher) Prefetch(block *types.Block, statedb *state.StateDB, c
 			return // Also invalid block, bail out
 		}
 		statedb.Prepare(tx.Hash(), i)
-		if err := precacheTransaction(msg, p.config, gaspool, statedb, header, evm); err != nil {
+
+		l1DataFee, err := fees.CalculateL1DataFee(tx, statedb)
+		if err != nil {
+			return
+		}
```
Comment (Thegaram) [inline core/state_prefetcher.go:77]: If the only possible error here is from the RLP encoding and if that will not happen in practice, then we could refactor these and panic instead of returning an error. Can be refactored in another PR.

File: core/types/l2trace.go [hunk 29]
```diff
@@ -44,7 +44,7 @@ type StorageTrace struct {
 // while replaying a transaction in debug mode as well as transaction
 // execution status, the amount of gas used and the return value
 type ExecutionResult struct {
-	L1Fee       *hexutil.Big `json:"l1Fee,omitempty"`
+	L1DataFee   *hexutil.Big `json:"l1DataFee,omitempty"`
```
Comment (Thegaram) [inline core/types/l2trace.go:47]: zkEVM can probably change when they change the type to `*hexutil.Big`. Can you please make sure that they know about this change?

File: eth/tracers/api_blocktrace.go [hunk 30]
```diff
@@ -258,7 +264,11 @@ func (api *API) getTxResult(env *traceEnv, state *state.StateDB, index int, bloc
 	state.Prepare(txctx.TxHash, txctx.TxIndex)
 
 	// Computes the new state by applying the given message.
-	result, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+	l1DataFee, err := fees.CalculateL1DataFee(tx, state)
+	if err != nil {
+		return err
```
Comment (Thegaram) [inline eth/tracers/api_blocktrace.go]:
```suggestion
		return fmt.Errorf("tracing failed: %w", err)
```

File: internal/ethapi/api.go [hunk 31]
```diff
@@ -1501,7 +1509,12 @@ func AccessList(ctx context.Context, b Backend, blockNrOrHash rpc.BlockNumberOrH
 		if err != nil {
 			return nil, 0, nil, err
 		}
-		res, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+		signer := types.MakeSigner(b.ChainConfig(), header.Number)
+		l1DataFee, err := fees.EstimateL1DataFeeForMessage(msg, header.BaseFee, b.ChainConfig().ChainID, signer, statedb)
+		if err != nil {
+			return nil, 0, nil, err
```
Comment (Thegaram) [inline internal/ethapi/api.go]:
```suggestion
			return nil, 0, nil, fmt.Errorf("failed to apply transaction: %v err: %v", args.toTransaction().Hash(), err)
```

File: les/state_accessor.go [hunk 32]
```diff
@@ -64,7 +65,11 @@ func (leth *LightEthereum) stateAtTransaction(ctx context.Context, block *types.
 		}
 		// Not yet the searched for transaction, execute on top of the current state
 		vmenv := vm.NewEVM(context, txContext, statedb, leth.blockchain.Config(), vm.Config{})
-		if _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(tx.Gas())); err != nil {
+		l1DataFee, err := fees.CalculateL1DataFee(tx, statedb)
+		if err != nil {
+			return nil, vm.BlockContext{}, nil, err
```
Comment (Thegaram) [inline les/state_accessor.go]:
```suggestion
			return nil, vm.BlockContext{}, nil, fmt.Errorf("transaction %#x failed: %v", tx.Hash(), err)
```

File: rollup/fees/rollup_fee.go [hunk 33]
```diff
@@ -48,59 +45,87 @@ type StateDB interface {
...
+	l1DataFee := calculateEncodedL1DataFee(raw, overhead, l1BaseFee, scalar)
+	return l1DataFee, nil
+}
+
+// asUnsignedTx turns a Message into a types.Transaction
+func asUnsignedTx(msg Message, baseFee, chainID *big.Int) *types.Transaction {
+	if baseFee == nil {
+		if msg.AccessList() == nil {
+			return asUnsignedLegacyTx(msg)
+		}
```
Comment (Thegaram) [inline rollup/fees/rollup_fee.go:75]: I think it's still possible that `type = AccessListTx` but `msg.AccessList() == nil`. In this case, it's possible that `EstimateL1DataFeeForMessage` will underestimate a bit, missing `ChainID` and the typed transaction envelope.

That said, the difference should be small and `EstimateL1DataFeeForMessage` is only used in APIs so we can keep it as it is.

**Review by Thegaram — DISMISSED** [review]

**Review by icemelon — DISMISSED** [review]

File: core/types/l2trace.go (hunk 29 above)
Comment (0xmountaintop) [reply core/types/l2trace.go:47]: I've informed them yesterday

File: rollup/fees/rollup_fee.go (hunk 33 above)
Comment (0xmountaintop) [reply rollup/fees/rollup_fee.go:75]: I agree.

One approach is we always overestimate a bit, by always do `AccessListTx` & `DynamicFeeTx`, but never do `LegacyTx`.

Another approach is, we can keep what we have now: we underestimate a bit in `asUnsignedTx`, as you analyze, then we overestimate in `WithSignature`.

both sound reasonable

File: les/state_accessor.go (hunk 32 above)
Comment (0xmountaintop) [reply les/state_accessor.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/a53292bb62432bff6fd8c5f0ba8c34bcdff4b572

File: internal/ethapi/api.go (hunk 31 above)
Comment (0xmountaintop) [reply internal/ethapi/api.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/a53292bb62432bff6fd8c5f0ba8c34bcdff4b572

File: eth/tracers/api_blocktrace.go (hunk 30 above)
Comment (0xmountaintop) [reply eth/tracers/api_blocktrace.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/a53292bb62432bff6fd8c5f0ba8c34bcdff4b572

File: core/rawdb/accessors_chain.go (hunk 27 above)
Comment (0xmountaintop) [reply core/rawdb/accessors_chain.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/507422f7750cf571a12cae00e2cfc0ad1bffe0c3
